File: frontend/src/lib/api/rollTables.ts
import type {
  RollTable,
  RollTableFormState,
  RollTableRollResult,
} from "../../features/campaigns/rollTableTypes";
import { request } from "./request";

function rollTablePayload(payload: RollTableFormState) {
  return {
    name: payload.name,
    description: payload.description,
    category: payload.category,
    tags: payload.tags
      .split(",")
      .map((tag) => tag.trim())
      .filter(Boolean),
    dieExpression: payload.dieExpression,
    rows: payload.rows.map((row) => ({
      minRoll: Number(row.minRoll) || 0,
      maxRoll: Number(row.maxRoll) || 0,
      label: row.label,
      resultText: row.resultText,
      notes: row.notes,
    })),
  };
}

export const rollTableApi = {
  campaignRollTables: (campaignId: string) =>
    request<{ tables: RollTable[] }>(`/api/campaigns/${campaignId}/roll-tables`),
  createCampaignRollTable: (campaignId: string, payload: RollTableFormState) =>
    request<{ table: RollTable }>(`/api/campaigns/${campaignId}/roll-tables`, {
      method: "POST",
      body: JSON.stringify(rollTablePayload(payload)),
    }),
  updateCampaignRollTable: (campaignId: string, tableId: string, payload: RollTableFormState) =>
    request<{ table: RollTable }>(`/api/campaigns/${campaignId}/roll-tables/${tableId}`, {
      method: "PUT",
      body: JSON.stringify(rollTablePayload(payload)),
    }),
  deleteCampaignRollTable: (campaignId: string, tableId: string) =>
    request<void>(`/api/campaigns/${campaignId}/roll-tables/${tableId}`, { method: "DELETE" }),
  cloneCampaignRollTable: (campaignId: string, tableId: string) =>
    request<{ table: RollTable }>(`/api/campaigns/${campaignId}/roll-tables/${tableId}/clone`, {
      method: "POST",
    }),
  rollCampaignRollTable: (campaignId: string, tableId: string) =>
    request<{ roll: RollTableRollResult }>(
      `/api/campaigns/${campaignId}/roll-tables/${tableId}/roll`,
      { method: "POST" },
    ),
};

File: backend/internal/models/roll_tables.go
package models

import "time"

type RollTable struct {
	ID            string         `json:"id"`
	CampaignID    string         `json:"campaignId,omitempty"`
	Source        string         `json:"source"`
	Name          string         `json:"name"`
	Description   string         `json:"description"`
	Category      string         `json:"category"`
	Tags          []string       `json:"tags"`
	DieExpression string         `json:"dieExpression"`
	Rows          []RollTableRow `json:"rows"`
	CreatedAt     time.Time      `json:"createdAt"`
	UpdatedAt     time.Time      `json:"updatedAt"`
}

type RollTableRow struct {
	ID         string `json:"id"`
	TableID    string `json:"tableId"`
	MinRoll    int    `json:"minRoll"`
	MaxRoll    int    `json:"maxRoll"`
	Label      string `json:"label"`
	ResultText string `json:"resultText"`
	Notes      string `json:"notes"`
	SortOrder  int    `json:"sortOrder"`
}

type RollTableRollResult struct {
	TableID       string    `json:"tableId"`
	TableName     string    `json:"tableName"`
	DieExpression string    `json:"dieExpression"`
	RolledValue   int       `json:"rolledValue"`
	MatchedRange  string    `json:"matchedRange"`
	Label         string    `json:"label"`
	ResultText    string    `json:"resultText"`
	Notes         string    `json:"notes"`
	RolledAt      time.Time `json:"rolledAt"`
}
